Remove simplification code from abstract value factories

hermanventer opened the pull request:
Rather than do some simplifications inside the factory functions, let the factory functions call simplifyAbstractValue. Some precautions are needed to avoid unbounded recursion.

NTillmann commented on src/realm.js:
```diff
@@ -232,6 +232,7 @@ export class Realm {
       metadata?: any
     ) => [Completion | Reference | Value, BabelNode, Array<BabelNodeStatement>],
   };
+  simplifyAbstractValue: (AbstractValue, boolean) => Value;
```
Mysterious boolean.
Is this function meant to be path sensitive?

hermanventer replied:
I'll document or remove.

NTillmann commented on src/utils/paths.js:
```diff
@@ -118,6 +118,6 @@ function pushInversePathCondition(condition: Value) {
 function pushRefinedConditions(unrefinedConditions: Array<AbstractValue>) {
   for (let unrefinedCond of unrefinedConditions) {
-    pushPathCondition(unrefinedCond.refineWithPathCondition());
+    pushPathCondition(unrefinedCond.$Realm.simplifyAbstractValue(unrefinedCond, true));
```
Mysterious `true` value.

NTillmann commented on src/utils/simplifier.js:
```diff
@@ -9,22 +9,23 @@
 import { AbstractObjectValue, AbstractValue, BooleanValue, ConcreteValue, Value } from "../values/index.js";
 
-export default function simplifyAbstractValue(realm: Realm, value: AbstractValue): Value {
+export default function simplifyAbstractValue(realm: Realm, value: AbstractValue, isCondition: boolean = false): Value {
```
I don't understand the intuition behind `isCondition`.

hermanventer replied:
x || true can simplify to just true if we don't care about the actual value of x, only its truthiness.

NTillmann commented on src/values/AbstractValue.js:
```diff
@@ -436,7 +359,10 @@ export default class AbstractValue extends Value {
 
     let resultTypes = TypesDomain.binaryOp(op, leftTypes, rightTypes);
-    let resultValues = ValuesDomain.binaryOp(realm, op, leftValues, rightValues);
+    let resultValues =
```
Why is this change part of an other behavior preserving pull request?

hermanventer replied:
The pull request is not entirely behavior preserving because a few simplifications done inside the factories are now replaced by full scale simplification. One of the effects of that is that templates now end up being asked for their truthiness and this breaks things. It is basically a bug that has been uncovered and it is convenient to fix it here.

hermanventer commented:
I've changed simplifyAbstractValue to simplifyAndRefineAbstractValue and introduced simplifyAndRefineAbstractCondition for expressions that are always converted to booleans. I have also made the factory for unary expressions call simplifyAbstractValue before returning its result. This uncovered a latent issue in a test case. I've modified the test case and referenced the relevant issue there.

NTillmann commented on src/realm.js:
```diff
@@ -232,7 +232,8 @@ export class Realm {
   };
-  simplifyAbstractValue: AbstractValue => Value;
+  simplifyAndRefineAbstractValue: AbstractValue => Value;
+  simplifyAndRefineAbstractCondition: AbstractValue => Value;
```
Just from the name of it, I still don't get what `...Condition` does. Is it path sensitive? Is it meant to only be called for particular kinds of conditions? Conditional expressions, or maybe binary conditions?

hermanventer replied:
Refinement is path sensitive by definition.

NTillmann replied:
I see. Now I understand :-) Can you persist that as a comment?

hermanventer replied:
Sure.
